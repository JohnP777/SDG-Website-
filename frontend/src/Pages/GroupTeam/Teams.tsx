import Page from '../../Components/Page';
import TeamsHeader from './TeamsHeader';
import TeamsList from './TeamsList';
import InvitationsList from './InvitationsList';
import React from 'react';

export interface TeamData {
  id: number,
  name: string,
  description: string,
  role?: string,
  invited_by?: string,
}

const Teams = () => {
  const [teams, setTeams] = React.useState<TeamData[]>([]);
  // List of teams must be accessible by both TeamsList and InvitationsList
  // so is supplied by parent container
  const addTeam = (team: TeamData) => {
    setTeams([...teams, team]);
  }

  return (
    <Page>
      <TeamsHeader />
      <TeamsList teams={teams} setTeams={setTeams} />
      <InvitationsList addTeam={addTeam} />
    </Page>
  );
  }
  
export default Teams;