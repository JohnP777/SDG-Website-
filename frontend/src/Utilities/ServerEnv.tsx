export const getUrl = () => {
    if (process.env.REACT_APP_BACKEND_URL) {
        return process.env.REACT_APP_BACKEND_URL;
    }
    return 'http://127.0.0.1:8000/'
}