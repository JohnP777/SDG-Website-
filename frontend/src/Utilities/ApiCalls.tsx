import { getUrl } from "./ServerEnv";

export const apiCallPost = async (path: string, body: any, isAuthed: boolean ) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'POST',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
      body: JSON.stringify(body)
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallGet = async (path: string, isAuthed: boolean, headersExtra?: any) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'GET',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : '',
        ...headersExtra
      },
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallPatch = async (path: string, body: object, isAuthed: boolean) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'PATCH',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
      body: JSON.stringify(body)
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallDelete = async (path: string, isAuthed: boolean) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'DELETE',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallPut = async (path: string, body: any, isAuthed: boolean) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'PUT',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
      body: JSON.stringify(body)
    });
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error);
  }
};
