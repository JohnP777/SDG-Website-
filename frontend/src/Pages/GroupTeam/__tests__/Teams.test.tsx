import { render, screen, fireEvent, waitFor, act } from "@testing-library/react";
import "@testing-library/jest-dom";
import * as ApiCalls from "../../../Utilities/ApiCalls";
import { MemoryRouter } from "react-router-dom";
import React from "react";
import Teams, { TeamData } from "../Teams";
import userEvent from "@testing-library/user-event";

// Mock the API calls
jest.mock("../../../Utilities/ApiCalls", () => ({
  apiCallGet: jest.fn(),
  apiCallPost: jest.fn(),
}));

// Mock window.location.reload
const mockReload = jest.fn();
Object.defineProperty(window, 'location', {
  value: {
    reload: mockReload,
  },
  writable: true
});

// Mock useNavigate
const mockedUsedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUsedNavigate,
}));

// Helper function to render the component with router
const renderWithRouter = (component: React.ReactNode) => {
  return render(
    <MemoryRouter>
      {component}
    </MemoryRouter>
  );
};

describe("TeamsList", () => {
  // Reset mocks before each test
  beforeEach(() => {
    jest.clearAllMocks();
    (ApiCalls.apiCallGet as jest.Mock).mockImplementation((url) => {
      if (url === 'api/auth/teams/') {
        return Promise.resolve({
          statusCode: 200,
          teams: [
            {
              id: 1,
              name: 'Team Green',
              description: '',
              role: 'owner',
            },
            {
              id: 2,
              name: 'Team Red',
              description: '',
              role: 'member',
            }
          ]
        })
      } else if (url === 'api/auth/teams/invitations/') {
        return Promise.resolve({
          statusCode: 200,
          pending_invitations: [
              {
                id: 3,
                name: 'Team Blue',
                description: '',
                role: 'member',
                invited_by: 'user1'
              },
              {
                id: 4,
                name: 'Team Yellow',
                description: '',
                role: 'member',
                invited_by: 'user1'
              }
          ]
        })
      } else if (url === 'api/teams/users/') {
        return Promise.resolve({
          statusCode: 200,
          usernames: ['user1', 'user2', 'user3'],
        })
      }
    });
    jest.clearAllMocks();
    (ApiCalls.apiCallPost as jest.Mock).mockImplementation((url) => {
      if (url === 'api/auth/teams/invitations/respond/') {
        return Promise.resolve({
          statusCode: 200,
        })
      }
    });
  });

  // Snapshot test
  it("should match snapshot", async () => {
    const { asFragment } = renderWithRouter(<Teams />);
    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });

  // Test TeamsList has headings
  it("should have teams headings", async () => {
    renderWithRouter(<Teams />);
    
    await waitFor(() => {
      expect(screen.getByText("Your Teams")).toBeInTheDocument();
      expect(screen.getByText("All your Teams!")).toBeInTheDocument();
    })
  });

  // Test TeamsList has team tiles
  it("should have team tiles", async () => {
    renderWithRouter(<Teams />);

    await waitFor(() => {
      expect(screen.getByText("Team Green")).toBeInTheDocument();
      expect(screen.getByText("Owner")).toBeInTheDocument();
      expect(screen.getByText("Team Red")).toBeInTheDocument();
      expect(screen.getByText("Member")).toBeInTheDocument();
    });
  });

  // Test TeamsList tiles navigate on click
  it("should navigate when team tile is clicked on", async () => {
    renderWithRouter(<Teams />);
    await waitFor(() => {
      const teamsTile = screen.getByText("Team Green");
      userEvent.click(teamsTile);
    })

    await waitFor(() => {
      expect(mockedUsedNavigate).toHaveBeenCalledWith('/teams/1')
    });
  });

  // Test InvitationsList has headings
  it("should have invitations headings", async () => {
    renderWithRouter(<Teams />);

    await waitFor(() => {
      expect(screen.getByText("Pending Invitations")).toBeInTheDocument();
      expect(screen.getByText("Awaiting your response")).toBeInTheDocument();
    });
  });

  // Test InvitationsList has team tiles
  it("should have invitations tiles", async () => {
    renderWithRouter(<Teams />);

    await waitFor(() => {
      expect(screen.getByText("Team Blue")).toBeInTheDocument();
      expect(screen.getByText("Team Yellow")).toBeInTheDocument();
      const subheadings = screen.getAllByText("New Invitation");
      expect(subheadings.length).toBe(2);
    });
  });

  // Test InvitationsList tiles open accept invite modal
  it("should have invite acceptance modal", async () => {
    renderWithRouter(<Teams />);

    await waitFor(() => {
      const invitationTile = screen.getByText("Team Blue");
      userEvent.click(invitationTile);
    })

    await waitFor(() => {
      // From mock data above: user 1 was the inviter
      expect(screen.getByText("user1")).toBeInTheDocument();
      expect(screen.getByText("has invited you to join their Team")).toBeInTheDocument();

      // From mock data above/team clicked on: team name is Team Blue
      // Team Blue should appear in the DOM twice: once in the list, and now a
      // second time in the modal.
      const teamNames = screen.getAllByText("Team Blue");
      expect(teamNames.length).toBe(2);

      // Interaction buttons
      expect(screen.getByText("Decline")).toBeInTheDocument();
      expect(screen.getByText("Accept Invitation")).toBeInTheDocument();
    });
  });

  // Test InvitationsList tiles decline button
  it("should delete invite if declined", async () => {
    renderWithRouter(<Teams />);

    await waitFor(() => {
      const invitationTile = screen.getByText("Team Blue");
      userEvent.click(invitationTile);
    })

    act(() => {
      const declineButton = screen.getByText("Decline");
      userEvent.click(declineButton);
    });

    await waitFor(() => {
      const invitationTile = screen.queryByText("Team Blue");
      expect(invitationTile).not.toBeInTheDocument();
    });
  });

  // Test InvitationsList tiles accept button
  it("should add team if accepted", async () => {
    renderWithRouter(<Teams />);
    await waitFor(() => {
      const invitationTile = screen.getByText("Team Blue");
      userEvent.click(invitationTile);
    })

    act(() => {
      const acceptButton = screen.getByText("Accept Invitation");
      userEvent.click(acceptButton);
    });

    await waitFor(() => {
      // Originally there were two invites
      // After accepting, there should only be one invite left
      const invitationTile = screen.getAllByText("New Invitation");
      expect(invitationTile.length).toBe(1);

      const acceptedTile = screen.getByText("Team Blue");
      expect(acceptedTile).toBeInTheDocument();

      // Originally there was only one team with Member role
      // Now there should be two
      const teamsSubheadings = screen.getAllByText("Member");
      expect(teamsSubheadings.length).toBe(2);
    });
  });
});