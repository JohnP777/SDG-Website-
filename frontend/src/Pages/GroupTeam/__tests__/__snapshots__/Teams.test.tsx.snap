// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamsList should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-z3339j"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-u2vtxm-MuiPaper-root-MuiAppBar-root"
      style="--Paper-shadow: none;"
    >
      <div
        class="MuiContainer-root css-12pwiy3-MuiContainer-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular css-1t0x8et-MuiToolbar-root"
        >
          <a
            class="MuiTypography-root MuiTypography-subtitle1 css-1z0p208-MuiTypography-root"
            href="/"
          >
            SDG
            <br />
            ZOO
          </a>
          <div
            class="MuiBox-root css-dq25au"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeLarge css-z77o6z-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1xf0eoi-MuiSvgIcon-root"
                data-testid="MenuIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 18h18v-2H3zm0-5h18v-2H3zm0-7v2h18V6z"
                />
              </svg>
            </button>
          </div>
          <div
            class="MuiBox-root css-10s323d"
          >
            <a
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1oyog3l-MuiButtonBase-root-MuiButton-root"
              href="/"
              tabindex="0"
            >
              HOME
            </a>
            <a
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1oyog3l-MuiButtonBase-root-MuiButton-root"
              href="/about-us"
              tabindex="0"
            >
              ABOUT
            </a>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-769nkz-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              FEATURES
              <span
                class="MuiButton-icon MuiButton-endIcon MuiButton-iconSizeMedium css-1gulhci-MuiButton-endIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="KeyboardArrowDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            class="MuiBox-root css-1t0e2o"
          >
            <button
              aria-label="Profile settings"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1avsmgh-MuiButtonBase-root-MuiIconButton-root"
              data-mui-internal-clone-element="true"
              tabindex="0"
              type="button"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-bqh3fr-MuiAvatar-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="PersonIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4m0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4"
                  />
                </svg>
              </div>
            </button>
          </div>
        </div>
      </div>
    </header>
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-eyrs0v-MuiContainer-root"
    >
      <div
        class="MuiBox-root css-69i1ev"
      >
        <div
          class="MuiBox-root css-0"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-f84ov1-MuiTypography-root"
          >
            Your Teams
          </h4>
          <p
            class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
          >
            All your Teams!
          </p>
        </div>
        <div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1317kya-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Create a Team
          </button>
        </div>
      </div>
      <div
        class="MuiBox-root css-fz0245"
      />
      <div
        class="MuiBox-root css-1m56bg2"
      />
    </div>
  </div>
</DocumentFragment>
`;
